tests/gdmem/malloc.c

```diff
@@ -0,0 +1,77 @@
+#include "gd.h"
+#include "gdtest.h"
+
+void *testcalloc(size_t nmemb, size_t size)
+{
+	char *ptr;
+
+	ptr = calloc(nmemb, size);
+
+	for (size_t i = 0; i < size; i++)
+		*(ptr + i) = 0;
+
+	return (void *)ptr;
+}
+
+void *testmalloc(size_t size)
+{
+	char *ptr;
+
+	ptr = malloc(size);
+
+	for (size_t i = 0; i < size; i++)
+		*(ptr + i) = 0;
+
+	return (void *)ptr;
+}
+
+void *testrealloc(void *ptr, size_t size)
+{
+	char *ret;
+
+	ret = realloc(ptr, size);
+
+	return (void *)ret;
+}
+
+void testfree(void *ptr)
+{
+	if (ptr != NULL)
+		free(ptr);
+}
+
+int main()
+{
+	gdImagePtr im1, im2;
+	FILE *fp;
+	CuTestImageResult result = {0, 0};
+
+	fp = gdTestFileOpen2("gdmem", "coin.png");
+	if (!gdTestAssert(fp != NULL))
+		return gdNumFailures();
+
+	im1 = gdImageCreateFromPng(fp);
+	if (!gdTestAssert(im1 != NULL))
+		goto door1;
+
+	gdSetMemoryCallocMethod(testcalloc);
+	gdSetMemoryMallocMethod(testmalloc);
+	gdSetMemoryReallocMethod(testrealloc);
+	gdSetMemoryFreeMethod(testfree);
+	fseek(fp, 0, SEEK_SET);
+	im2 = gdImageCreateFromPng(fp);
+	if (!gdTestAssert(im2 != NULL))
+		goto door2;
+
+	gdTestImageDiff(im1, im2, NULL, &result);
+	gdTestAssertMsg(result.pixels_changed == 0, "pixels changed: %d\n", result.pixels_changed);
```

Comment:
this doesn't really test that the allocators worked, just that the methods above didn't crash.  maybe a better test would be to have each test func above increment a global variable and then we assert that ?

```
static int free_called = 0;
void testfree(void *ptr)
{
    free_called++;
    free(ptr);
}
...
	im1 = gdImageCreateFromPng(fp);
	if (!gdTestAssert(im1 != NULL))
		goto door1;

        ... assert that free_called == 0 and all other funcs too ...

	gdSetMemoryCallocMethod(testcalloc);
	gdSetMemoryMallocMethod(testmalloc);
	gdSetMemoryReallocMethod(testrealloc);
	gdSetMemoryFreeMethod(testfree);
...
        ... assert that free_called != 0 and all other funcs too ...
```

Reply:
Done much like this, but instead of counting the calls I went with only checking if it was called.